src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs

```diff
@@ -16,18 +16,257 @@
+                }
+                else if (this.IsExplicitInterfaceImplementation)
+                {
+                    typeMap = new ExplicitInterfaceMethodTypeParameterMap(this);
+                }
+
+                var typeParameters = syntax.TypeParameterList.Parameters;
+                var result = ArrayBuilder<TypeParameterSymbol>.GetInstance();
+
+                for (int ordinal = 0; ordinal < typeParameters.Count; ordinal++)
+                {
+                    var parameter = typeParameters[ordinal];
+                    if (parameter.VarianceKeyword.Kind() != SyntaxKind.None)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_IllegalVarianceSyntax, parameter.VarianceKeyword.GetLocation());
+                    }
+
+                    var identifier = parameter.Identifier;
+                    var location = identifier.GetLocation();
+                    var name = identifier.ValueText;
+
+                    // Note: It is not an error to have a type parameter named the same as its enclosing method: void M<M>() {}
+
+                    for (int i = 0; i < result.Count; i++)
+                    {
+                        if (name == result[i].Name)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_DuplicateTypeParameter, location, name);
+                            break;
+                        }
+                    }
+
+                    SourceMemberContainerTypeSymbol.ReportReservedTypeName(identifier.Text, this.DeclaringCompilation, diagnostics.DiagnosticBag, location);
+
+                    var tpEnclosing = ContainingType.FindEnclosingTypeParameter(name);
+                    if ((object)tpEnclosing != null)
+                    {
+                        // Type parameter '{0}' has the same name as the type parameter from outer type '{1}'
+                        diagnostics.Add(ErrorCode.WRN_TypeParameterSameAsOuterTypeParameter, location, name, tpEnclosing.ContainingType);
+                    }
+
+                    var syntaxRefs = ImmutableArray.Create(parameter.GetReference());
+                    var locations = ImmutableArray.Create(location);
+                    var typeParameter = (typeMap != null) ?
+                        (TypeParameterSymbol)new SourceOverridingMethodTypeParameterSymbol(
+                            typeMap,
+                            name,
+                            ordinal,
+                            locations,
+                            syntaxRefs) :
+                        new SourceMethodTypeParameterSymbol(
+                            this,
+                            name,
+                            ordinal,
+                            locations,
+                            syntaxRefs);
+
+                    result.Add(typeParameter);
+                }
+
+                return result.ToImmutableAndFree();
+            }
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

It is extremely confusing when methods of an enclosing type continue after quite large nested type. Consider moving nested types to the bottom of the enclosing type body.

Reply:
Done.  I'm also ok with these being extracted to separate partial files.  I wasn't certain if tehre was a preference wrt to the review to make it (possibly) slightly easier to understand.